limacharlie/retention.go

```diff
@@ -89,3 +89,102 @@ func (org *Organization) EventByAtom(sensorID, atom string) (EventContainer, err
+	// End is the required timestamp in seconds where to stop fetching detections
+	End int `json:"end"`
+	// Limit maximum number of detections to return
+	Limit int `json:"limit"`
+}
+
+func (org Organization) HistoricalDetections(detectionReq HistoricalDetectionsRequest) (HistoricalDetectionsResponse, error) {
+
+	var results HistoricalDetectionsResponse
+
+	if detectionReq.Cursor == "" {
+		detectionReq.Cursor = "-"
+	}
+
+	q := makeDefaultRequest(&results)
+	q = q.withQueryData(Dict{
+		"cat":    detectionReq.Cat,
+		"cursor": detectionReq.Cursor,
+		"start":  detectionReq.Start,
+		"end":    detectionReq.End,
+		"limit":  detectionReq.Limit,
+	})
+
+	if err := org.client.reliableRequest(http.MethodGet, fmt.Sprintf("insight/%s/detections", org.client.options.OID), q); err != nil {
+		return results, err
+	}
+
+	return results, nil
+}
+
+type HistoricalDetectionsResponse struct {
+	Detects    []Detects `json:"detects"`
+	NextCursor string    `json:"next_cursor"`
+}
+
+type Detect struct {
+	Event   Event   `json:"event"`
+	Routing Routing `json:"routing"`
+}
+type DetectMetadata struct {
+	Author         string   `json:"author"`
+	Description    string   `json:"description"`
+	Falsepositives []string `json:"falsepositives"`
+	Level          string   `json:"level"`
+	References     []string `json:"references"`
+	Tags           []string `json:"tags"`
+}
+
+type Detects struct {
+	Author    string         `json:"author"`
+	Cat       string         `json:"cat"`
+	Detect    Detect         `json:"detect"`
+	DetectID  string         `json:"detect_id"`
+	DetectMtd DetectMetadata `json:"detect_mtd"`
+	Link      string         `json:"link"`
+	Namespace string         `json:"namespace"`
+	Routing   Routing        `json:"routing"`
+	Source    string         `json:"source"`
+	Ts        int64          `json:"ts"`
+}
+
+const NewProcessEventType = "NEW_PROCESS"
+
+type NewProcessEvent struct {
+	BaseAddress     int64  `json:"BASE_ADDRESS"`
+	CommandLine     string `json:"COMMAND_LINE"`
+	FileIsSigned    int    `json:"FILE_IS_SIGNED"`
+	FilePath        string `json:"FILE_PATH"`
+	Hash            string `json:"HASH"`
+	MemoryUsage     int    `json:"MEMORY_USAGE"`
+	Parent          Parent `json:"PARENT"`
+	ParentProcessID int    `json:"PARENT_PROCESS_ID"`
+	ProcessID       int    `json:"PROCESS_ID"`
+	Threads         int    `json:"THREADS"`
+	UserName        string `json:"USER_NAME"`
+}
+
+type Parent struct {
+	CommandLine     string `json:"COMMAND_LINE"`
+	FileIsSigned    int    `json:"FILE_IS_SIGNED"`
+	FilePath        string `json:"FILE_PATH"`
+	Hash            string `json:"HASH"`
+	MemoryUsage     int    `json:"MEMORY_USAGE"`
+	ParentAtom      string `json:"PARENT_ATOM"`
+	ParentProcessID int    `json:"PARENT_PROCESS_ID"`
+	ProcessID       int    `json:"PROCESS_ID"`
+	ThisAtom        string `json:"THIS_ATOM"`
+	Threads         int    `json:"THREADS"`
+	Timestamp       int64  `json:"TIMESTAMP"`
+	UserName        string `json:"USER_NAME"`
+}
```

Comment:
Since there's no specific schema for detections and events, it's difficult to make struct globally like this. I suggest you bring those definitions to your user-code where you might know ahead of time what kind of detection and events you're going to get.
```suggestion
```

Reply:
Makes sense, It'll be better to keep it generic in the library. I'll implement the event types and event in my own code